mono/metadata/icall.c

```diff
@@ -2163,25 +2163,39 @@ ves_icall_MonoField_SetValueInternal (MonoReflectionFieldHandle field, MonoObjec
 		mono_gchandle_free (value_gchandle);
 }
 
+ICALL_EXPORT MonoObject*
+mono_TypedReference_ToObject (MonoTypedRef* tref);
+
 ICALL_EXPORT void
 ves_icall_System_RuntimeFieldHandle_SetValueDirect (MonoReflectionField *field, MonoReflectionType *field_type, MonoTypedRef *obj, MonoObject *value, MonoReflectionType *context_type)
 {
+	HANDLE_FUNCTION_ENTER ();
+
+	ERROR_DECL (error);
+
 	MonoClassField *f;
+	MonoObjectHandle objHandle, valueHandle;
+	MonoReflectionFieldHandle fieldHandle;
 
 	g_assert (field);
 	g_assert (obj);
 	g_assert (value);
 
 	f = field->field;
 	if (!MONO_TYPE_ISSTRUCT (m_class_get_byval_arg (f->parent))) {
-		mono_set_pending_exception (mono_get_exception_not_implemented (NULL));
-		return;
+		fieldHandle = MONO_HANDLE_NEW (MonoReflectionField, field);
+		objHandle = MONO_HANDLE_NEW (MonoObject, mono_TypedReference_ToObject (obj));
+		valueHandle = MONO_HANDLE_NEW (MonoObject, value);
+		ves_icall_MonoField_SetValueInternal (fieldHandle, objHandle, valueHandle, error);
+		mono_error_set_pending_exception (error);
```

Comment:
better to pair this up with `HANDLE_FUNCTION_RETURN()`.

Even better would be to write the function like this:
```c
ICALL_EXPORT void
 ves_icall_System_RuntimeFieldHandle_SetValueDirect (MonoReflectionField *field, MonoReflectionType *field_type, MonoTypedRef *obj, MonoObject *value, MonoReflectionType *context_type)
{
	ICALL_ENTRY();
	/* "error" already declared by ICALL_ENTRY, HANDLE_FUNCTION_ENTRY also called. */
	...
	/* mono_error_set_pending_exception already called by ICALL_RETURN, HANDLE_FUNCTION_RETURN also called. */
	ICALL_RETURN();
}
```

Reply:
I never used ICALL_ENTRY/RETURN. Is it worth another pattern?